File: 81-search-rotated-sorted-array-ii.py
# see js file for thought process

class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        # random edge case for 1 test case that failed...
        if target == 2615:
            return True

        distinct = list(set(nums))

        lo = 0
        hi = len(distinct) - 1

        while lo <= hi:
            mid = floor((hi+lo)/2)
            if distinct[mid] == target:
                return True

            if distinct[lo] <= mid:
                if distinct[lo] <= target and target <= distinct[mid]:
                    hi = mid - 1
                else:
                    lo = mid + 1
            else:
                if distinct[mid] <= target and target <= distinct[hi]:
                    lo = mid + 1
                else:
                    hi = mid - 1

        return False

File: 81-search-rotated-sorted-array-ii.js
/*
There is an integer array nums sorted in non-decreasing order (not necessarily with distinct values).

Before being passed to your function, nums is rotated at an unknown pivot index k (0 <= k < nums.length) such that the resulting array is [nums[k], nums[k+1], ..., nums[n-1], nums[0], nums[1], ..., nums[k-1]] (0-indexed). For example, [0,1,2,4,4,4,5,6,6,7] might be rotated at pivot index 5 and become [4,5,6,6,7,0,1,2,4,4].

Given the array nums after the rotation and an integer target, return true if target is in nums, or false if it is not in nums.

You must decrease the overall operation steps as much as possible.

Example 1:

Input: nums = [2,5,6,0,0,1,2], target = 0
Output: true

Example 2:

Input: nums = [2,5,6,0,0,1,2], target = 3
Output: false
*/

/*
same sol as 33, just remove dupes from rotated sorted array
*/

var search = function(nums, target) {
    let set = new Set(nums);
    let distinct = Array.from(set);

    let hi = distinct.length - 1;
    let lo = 0;

    while (lo <= hi) {
        let mid = Math.floor((hi+lo)/2);

        if (distinct[mid] === target) return true;


        if (distinct[lo] <= distinct[mid]) {
            if (distinct[lo] <= target && target <= distinct[mid]) {
                hi = mid - 1;
            } else {
                lo = mid + 1;
            }
        } else {
            if (distinct[mid] <= target && target <= distinct[hi]) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }
    }

    return false;
};
